### ISSUE TITLE
tf.image.ssim_multiscale has a bug as a loss function. It produces nan. The bug didn't happen before. Please fix it

### ISSUE THREAD

**Issue by yifeipet** [tensorflow/tensorflow#57493]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.9.1

### Custom Code

Yes

### OS Platform and Distribution

Windows

### Python version

3.8

### CUDA/cuDNN version

11.2/8.1

### GPU model and memory

RTX A6000

### Current Behaviour?

```shell
tf.image.ssim_multiscale has a bug as a loss function. It produces nan during training. The bug didn't happen in the previous Tensorflow versions. I think another person had a similar issue with the ssim loss.  Colab Tensorflow version is 2.8.2 and it doesn't have this issue. But I really need to use Tensorflow 2.9.1 for my project. Please help us to fix it. Thank you!
```

### Standalone code to reproduce the issue

```shell
msssim = tf.reduce_mean(tf.image.ssim_multiscale(real_images, generated_images, 2))
```

### Relevant log output

```shell
Epoch 1/10
   6/7856 [..............................] - ETA: 1:45:39 - d_loss: nan - g_loss: nan - ms_ssim: nan - l1: nan - l2: nan - entropy: nan - g_cost: nan
```
</details>

**Comment by bhack** [comment]
As you don't have a minimal example/dummy network to reproduce it can you check if it could be the same as https://github.com/tensorflow/tensorflow/issues/50400?

**Comment by yifeipet** [comment]
> As you don't have a minimal example/dummy network to reproduce it can you check if it could be the same as #50400?

I have already uninstalled Tensorflow 2.9.1 and reinstalled Tensorflow 2.4.0. The MS-SSIM loss in my customized gradient tape like the code sample you mentioned produces Nan in Tensorflow 2.9.1 and Tensorflow 2.8.2 on my RTX A6000 workstation, but not in Tensorflow 2.4.0 on my RTX A6000 workstation. But current Colab Tensorflow 2.8.2 doesn't have the error for the same codes. I don't know the reason. Is the current Tensorflow not friendly with RTX A6000?  Please help me out and solve the issue in the future version of Tensorflow. Otherwise, I cannot use a newer Tensorflow version in the future. Thank you!

**Comment by bhack** [comment]
Please check https://github.com/tensorflow/tensorflow/issues/50400#issuecomment-1231502107.

**Comment by sushreebarsa** [comment]
@yifeipet Could you please refer the [comment](https://github.com/tensorflow/tensorflow/issues/57493#issuecomment-1231502421) above?
Thank you!

**Comment by yifeipet** [comment]
> @yifeipet Could you please refer the [comment](https://github.com/tensorflow/tensorflow/issues/57493#issuecomment-1231502421) above? Thank you!

But why can Tensorflow 2.4 do not have this issue? You cannot just throw me a paper and say I cannot do anything. You should check why Tensorflow 2.4 doesn't have this issue. I am very unsatisfied with Tensorflow now. It wastes me a lot of time on the bug that hasn't been on previous Tensorflow versions.

**Comment by bhack** [comment]
> and indicate I cannot do anything

Who is claiming this? I have proposed one popular solution in that comment that it is adopted in many implementations when you are going to use this metrics in a differentiable way as loss.

>  Please let me know where I can find the original Tensorflow 2.4 source codes for SSIM.

You can see the blame date to check eventually the most recent changes in the implementation but I don't think you will find something interesting: 
https://github.com/tensorflow/tensorflow/blame/master/tensorflow/python/ops/image_ops_impl.py

Then as I don't see you code, your random seeds etc.. it is hard to reproduce your issue across TF versions, 
But generally just using SSIM or MS-SSIM metrics, as is, in a loss it is not safe without some additional care in you loss building.

**Comment by yifeipet** [comment]
> > and indicate I cannot do anything
> 
> Who is claiming this? I have proposed one popular solution in that comment that it is adopted in many implementations when you are going to use this metrics in a differentiable way as loss.
> 
> > Please let me know where I can find the original Tensorflow 2.4 source codes for SSIM.
> 
> You can see the blame date to check eventually the most recent changes in the implementation but I don't think you will find something interesting: https://github.com/tensorflow/tensorflow/blame/master/tensorflow/python/ops/image_ops_impl.py
> 
> Then as I don't see you code, your random seeds etc.. it is hard to reproduce your issue across TF versions, But generally just using SSIM or MS-SSIM metrics, as is, in a loss it is not safe without some additional care in you loss building.

Sorry for that. But it took me 3 days to try multiple ways to install Tensorflow and uninstall it multiple times in order to find SSIM bug in Tensorflow, which is quite annoying. Please understand it. I might try to clip the SSIM value when I use a newer Tensorflow version. However, I still hope this issue solved in the future update.

**Comment by bhack** [comment]
@yifeipet Currently we are not offering a SSIM loss in TF directly but just an operator. 
I suggest to request or contribute a SSIM/MS-SSIM loss in https://github.com/keras-team/keras-cv if they are interested. 
It could include any extra guard in the loss implementation itself.

**Comment by yifeipet** [comment]
> @yifeipet Currently we are not offering a SSIM loss in TF directly but just an operator. I suggest to request or contribute a SSIM/MS-SSIM loss in https://github.com/keras-team/keras-cv if they are interested. It could include any extra guard in the loss implementation itself.

Thank you! Now I just need to use Tensorflow 2.4 for my project as the deadline is very close. Then if I have time to try newer Tensorflow, I will try the clip value method by myself to see if it can solve the issue in a newer Tensorflow version. Thanks for your help.

**Comment by gadagashwini** [comment]
Hi @yifeipet, Did you get a chance to work on Latest release Tensorflow version. Thank you!

**Comment by yifeipet** [comment]
> Hi @yifeipet, Did you get a chance to work on Latest release Tensorflow version. Thank you!

I am still using Tensorflow 2.4 for me current project. The deadline is October 24. Then, I might try new version. But anyway, thank you!

**Comment by tilakrayal** [comment]
@yifeipet,
I tried to execute with the sample code on the TensorFlow v2.17 and it was executed without any issues. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/0a40c50e4a49bafb7d8f3001da02c60a/untitled2138.ipynb). Also the Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue.

Thank you!
